src/v/cloud_storage_clients/s3_client.cc

```diff
@@ -561,11 +563,102 @@ s3_client::s3_client(
+          "Can't self-configure S3 Client, {}, {} are not enabled. Defaulting "
+          "to {}",
+          config::shard_local_cfg().cloud_storage_enable_remote_read.name(),
+          config::shard_local_cfg().cloud_storage_enable_remote_write.name(),
+          result.url_style);
+        co_return result;
+    }
+
+    const auto& bucket_config = config::shard_local_cfg().cloud_storage_bucket;
+
+    if (!bucket_config.value().has_value()) {
+        vlog(
+          s3_log.warn,
+          "Can't self-configure S3 Client, {} is not set. Defaulting to {}",
+          bucket_config.name(),
+          result.url_style);
+        co_return result;
+    }
+
+    const auto bucket = cloud_storage_clients::bucket_name{
+      bucket_config.value().value()};
+
+    // Test virtual_host style.
+    vassert(
+      _requestor._ap_style == s3_url_style::virtual_host,
+      "_ap_style should be virtual host by default before self configuration "
+      "begins");
+    if (co_await self_configure_test(bucket, remote_read, remote_write)) {
+        // Virtual-host style request succeeded.
+        co_return result;
+    }
+
+    // Test path style.
+    _requestor._ap_style = s3_url_style::path;
+    result.url_style = _requestor._ap_style;
+    if (co_await self_configure_test(bucket, remote_read, remote_write)) {
+        // Path style request succeeded.
+        co_return result;
+    }
+
+    // Both addressing styles failed.
     vlog(
       s3_log.error,
-      "Call to self_configure was made, but the S3 client doesn't require self "
-      "configuration");
-    co_return s3_self_configuration_result{};
+      "Couldn't reach S3 storage with either path style or virtual_host style "
+      "requests.",
+      bucket_config.name());
+    co_return error_outcome::fail;
+}
+
+ss::future<bool> s3_client::self_configure_test(
+  const bucket_name& bucket, bool remote_read, bool remote_write) {
+    if (remote_read) {
+        // Verify with a list objects request.
+        auto list_objects_result = co_await list_objects(
+          bucket, std::nullopt, std::nullopt, 1);
+        co_return list_objects_result;
+    } else {
+        vassert(remote_write, "Remote write is not enabled");
+        // Verify with a upload and delete request.
+        const ss::sstring key_and_payload = "S3ClientSelfConfigurationKey";
+        iobuf payload;
+        payload.append(key_and_payload.data(), key_and_payload.size());
+        auto payload_stream = make_iobuf_input_stream(std::move(payload));
+        const ss::lowres_clock::duration timeout = {
+          std::chrono::milliseconds(1000)};
+
+        auto upload_object_result = co_await put_object(
+          bucket,
+          object_key{key_and_payload},
+          key_and_payload.size(),
+          std::move(payload_stream),
+          timeout);
+
+        if (!upload_object_result) {
+            // Upload failed, return early.
+            co_return upload_object_result;
+        }
```

Comment:
Just curious, is there a specific error that is returned when using the wrong path style?

Reply:
No, I believe any error returned would be some generic HTTP response code.

Reply:
Is there at least a way to differentiate between this and a timeout? My concern is that two timeouts will cause Redpanda to abort, which makes picking a correct timeout pretty important.

Reply:
What sort of behaviour do you think the case with two timeouts should have, and does it differ from two error responses?

IMO, two error codes should lead to termination, as should two timeouts (since this is the self-configuration step, I don't see why timeouts shouldn't be treated different from errors), but I am interested in how you think the error-handling should behave here.

Reply:
I could imagine a timeout just being retried on for some period, though I suppose that would be roughly equivalent to starting with a longer timeout.

Also I wonder if it's reasonable to just fall back to virtual host, given the alternative will be deprecated soon.

In any case, I'm more comfortable with this as is with a larger timeout.